summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/AsyncBase.scala

```diff
@@ -106,15 +135,18 @@ abstract class AsyncBase[I, O, S, D, RC](maxWaitingFutures: MaxWaitingFutures, m
     val pendingFuturesCount = numPendingOutstandingFutures.get
     if (pendingFuturesCount > maxWaitingFuturesCount) {
       // Too many futures waiting, let's clear.
-      outstandingFutures.dequeueAll(_.isDefined)
-      val toForce = outstandingFutures.trimTo(maxWaitingFuturesCount).toIndexedSeq
-      if (toForce.nonEmpty) {
+      val pending = outstandingFutures.toSeq.filterNot(_.isDefined)
+      val toClear = pending.size - maxWaitingFuturesCount
+      if (toClear > 0) {
         try {
-          Await.ready(Future.collect(toForce), maxWaitingTime.get)
+          Await.ready(AsyncBase.waitN(pending, toClear), maxWaitingTime.get)
+          outstandingFutures.putAll(pending.filterNot(_.isDefined))
```

Comment:
I'm worried that is the `TimeoutException` happens, then we never put any futures back. I think in that case we should put them all back? What do you think about?

```scala
try {
   Await.ready(AsyncBase.waitN(pending, toClear), maxWaitingTime.get)
} catch {
  case te: TimeoutException =>
}
outstandingFutures.putAll(pending.filterNot(_.isDefined))
```